### PR TITLE
Add simple where condition to filter on indexed fields

### PR DESCRIPTION
**Pull request by BenoitLacour**
Hello,

This is a proposal to add a simple where filter to filter on indexed fields. I needed it to make an app I work on faster to recover entities linked data.

Maybe I'm wrong and it has already another solution, but didn't find it so this PR could help other users.

Best regards,
Benoît

### CODE REVIEW

File: src/BlazorDB/wwwroot/blazorDB.js [hunk 1]
```diff
@@ -132,6 +132,21 @@ window.blazorDB = {
             });
         });
     },
+    where: function(dotnetReference, transaction, dbName, storeName, condition) {
+        return new Promise((resolve, reject) => {
+            window.blazorDB.getDb(dbName).then(db => {
+                db[storeName].where(JSON.parse(condition)).toArray(items => {
```
Comment (nwestfall) [inline src/BlazorDB/wwwroot/blazorDB.js:138]: Can you use this approach instead?  Just to be consistent

https://github.com/nwestfall/BlazorDB/blob/1ef4e14d8eea7ce0f47e7430eae63fa23f6aee44/src/BlazorDB/wwwroot/blazorDB.js#L123

File: example/BlazorDB.Example/wwwroot/index.html [hunk 2]
```diff
@@ -1,26 +1,26 @@
...
+    <app>Loading...</app>
+
+    <div id="blazor-error-ui">
+        An unhandled error has occurred.
+        <a href="" class="reload">Reload</a>
+        <a class="dismiss">🗙</a>
+    </div>
+    <script src="_framework/blazor.webassembly.js"></script>
+    <script src="_content/BlazorIndexedDB/dexie.min.js"></script>
+    <script src="_content/BlazorIndexedDB/blazorDB.js"></script>
```
Comment (nwestfall) [inline example/BlazorDB.Example/wwwroot/index.html:23]: I don't think this folder should change, but I'll pull locally to test

File: src/BlazorDB/IndexDbManager.cs [hunk 3]
```diff
@@ -300,7 +299,41 @@ public async Task<IList<TRecord>> ToArray<TRecord>(string storeName)
...
+
+        private string FormatConditionJsonString(string key, object value)
+        {
+            if (value is string)
+            {
+                return $"{{ \"{key}\": \"{value}\" }}";
+            }
+            
+            return $"{{ \"{key}\": {value} }}";
+        }
```
Comment (nwestfall) [inline src/BlazorDB/IndexDbManager.cs:335]: I have to think about this, the condition for `is string` is good, but if the `object` is not something like an `int`, we might want to serialize it.

Or, you pass the key/value to Javascript and have that worry about the serialization

Comment (nwestfall) [inline src/BlazorDB/IndexDbManager.cs:335]: I have to think about this, the condition for `is string` is good, but if the `object` is not something like an `int`, we might want to serialize it.

Or, you pass the key/value to Javascript and have that worry about the serialization

**Review by nwestfall — CHANGES_REQUESTED** [review]
Some of it I have to test, but let me know your thoughts.  Love the contribution!

File: example/BlazorDB.Example/wwwroot/index.html (hunk 2 above)
Comment (BenoitLacour) [reply example/BlazorDB.Example/wwwroot/index.html:23]: It was not working with the previous value on my computer, I was thinking it was an unseen regression, so I changed it to make it works on my side. So if it works from your side, there is something I have to understand...

File: src/BlazorDB/wwwroot/blazorDB.js (hunk 1 above)
Comment (BenoitLacour) [reply src/BlazorDB/wwwroot/blazorDB.js:138]: I will take a look at it.

File: src/BlazorDB/IndexDbManager.cs (hunk 3 above)
Comment (BenoitLacour) [reply src/BlazorDB/IndexDbManager.cs:335]: Yeah, I'm not very confident about this code, the thing is we have to generate some JSON to be interpreted on javascript side, the keys are dynamic, so it was "easier" to generate it as a simple string on c# side.

I will dig about this subject a bit.

Comment (BenoitLacour) [reply src/BlazorDB/IndexDbManager.cs]: @nwestfall I changed the way the filter object is generated, delegating the creation of the json to javascript part, this put the responsability to form the correct filter on js side, closer to dexie.
In an other way, we let c# serialize the IndexFilterValue, so we don't care who need quotes or not anymore.

Let me know what you think about it.

**Comment by BenoitLacour** [comment]
@nwestfall Did you see my updates?

**Comment by etaxi341** [comment]
Any news on this?

File: example/BlazorDB.Example/wwwroot/index.html (hunk 2 above)
Comment (etaxi341) [reply example/BlazorDB.Example/wwwroot/index.html:23]: > It was not working with the previous value on my computer, I was thinking it was an unseen regression, so I changed it to make it works on my side. So if it works from your side, there is something I have to understand...

For me it only works with your change too. It looks like "BlazorIndexedDB" is the correct path

**Comment by nwestfall** [comment]
Sorry for the delay guys, had a few other projects come up.  I'm pulling the code and taking a look to try and get this out.

File: src/BlazorDB/IndexDbManager.cs [hunk 4]
```diff
@@ -300,7 +300,43 @@ public async Task<IList<TRecord>> ToArray<TRecord>(string storeName)
...
+        /// <summary>
+        /// Filter a store on an indexed value 
+        /// </summary>
+        /// <param name="storeName">The name of the store to retrieve the records from</param>
+        /// <param name="indexName">index field name to filter on</param>
+        /// <param name="filterValue">filter's value</param>
+        /// <returns></returns>
+        public async Task<IList<TRecord>> Where<TRecord>(string storeName, string indexName, object filterValue)
+        {
+            var filters = new List<IndexFilterValue>() {new(indexName, filterValue)};
```
Comment (nwestfall) [inline src/BlazorDB/IndexDbManager.cs:313]: What version of C# is this?  I don't think I'm using it yet, but I'm always happy to update this.

**Review by nwestfall — CHANGES_REQUESTED** [review]
I think just fix the build and this will be good to go.  Like the approach you took for the`IndexFilterValue`

**Comment by nwestfall** [comment]
@BenoitLacour never mind, I did not mean to merge this, not sure what happened.  But now that it's in master, I can go ahead and fix it.  Thanks!

File: src/BlazorDB/IndexDbManager.cs (hunk 4 above)
Comment (BenoitLacour) [reply src/BlazorDB/IndexDbManager.cs:313]: It's a C# 9.0 feature, maybe not ideal to use on a project that target .Net Core 3.1 but it is compatible with.

**Pull request merged**